**[Machine Learning Micro-Course Home Page](https://www.kaggle.com/learn/machine-learning)**

---


# Introduction
Machine learning competitions are a great way to improve your data science skills and measure your progress. 

In this exercise, you will create and submit predictions for a Kaggle competition. You can then improve your model (e.g. by adding features) to improve and see how you stack up to others taking this micro-course.

The steps in this notebook are:
1. Build a Random Forest model with all of your data (**X** and **y**)
2. Read in the "test" data, which doesn't include values for the target.  Predict home values in the test data with your Random Forest model.
3. Submit those predictions to the competition and see your score.
4. Optionally, come back to see if you can improve your model by adding features or changing your model. Then you can resubmit to see how that stacks up on the competition leaderboard.

## Recap
Here's the code you've written so far. Start by running it again.

In [1]:
# Code you have previously used to load data
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from learntools.core import *



# Path of the file to read. We changed the directory structure to simplify submitting to a competition
iowa_file_path = '../input/train.csv'

home_data = pd.read_csv(iowa_file_path)
# Create target object and call it y
y = home_data.SalePrice
# Create X
features = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']
X = home_data[features]

# Split into validation and training data
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state=1)

# Specify Model
iowa_model = DecisionTreeRegressor(random_state=1)
# Fit Model
iowa_model.fit(train_X, train_y)

# Make validation predictions and calculate mean absolute error
val_predictions = iowa_model.predict(val_X)
val_mae = mean_absolute_error(val_predictions, val_y)
print("Validation MAE when not specifying max_leaf_nodes: {:,.0f}".format(val_mae))

# Using best value for max_leaf_nodes
iowa_model = DecisionTreeRegressor(max_leaf_nodes=100, random_state=1)
iowa_model.fit(train_X, train_y)
val_predictions = iowa_model.predict(val_X)
val_mae = mean_absolute_error(val_predictions, val_y)
print("Validation MAE for best value of max_leaf_nodes: {:,.0f}".format(val_mae))

# Define the model. Set random_state to 1
rf_model = RandomForestRegressor(random_state=1)
rf_model.fit(train_X, train_y)
rf_val_predictions = rf_model.predict(val_X)
rf_val_mae = mean_absolute_error(rf_val_predictions, val_y)

print("Validation MAE for Random Forest Model: {:,.0f}".format(rf_val_mae))


Validation MAE when not specifying max_leaf_nodes: 29,653
Validation MAE for best value of max_leaf_nodes: 27,283
Validation MAE for Random Forest Model: 22,762


/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


# Creating a Model For the Competition

Build a Random Forest model and train it on all of **X** and **y**.  

In [2]:
# To improve accuracy, create a new Random Forest model which you will train on all training data
rf_model_on_full_data = RandomForestRegressor(random_state=1)

# fit rf_model_on_full_data on all data from the training data
rf_model_on_full_data.fit(X,y)


/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
           oob_score=False, random_state=1, verbose=0, warm_start=False)

# Make Predictions
Read the file of "test" data. And apply your model to make predictions

In [3]:
# path to file you will use for predictions
test_data_path = '../input/test.csv'

# read test data file using pandas
test_data = pd.read_csv(test_data_path)

# create test_X which comes from test_data but includes only the columns you used for prediction.
# The list of columns is stored in a variable called features
test_X = test_data[features]

# make predictions which we will submit. 
test_preds = rf_model_on_full_data.predict(test_X)

# The lines below shows how to save predictions in format used for competition scoring
# Just uncomment them.

#output = pd.DataFrame({'Id': test_data.Id,
#                       'SalePrice': test_preds})
#output.to_csv('submission.csv', index=False)

# Improve Model: Trial 1

In [4]:
# see list of all features
print(f"Features used in initial model: {features}\n")
print(home_data.columns.tolist())

Features used in initial model: ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']

['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating', 'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual', 'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQua

In [5]:
home_data.head()

Id  MSSubClass MSZoning    ...     SaleType  SaleCondition SalePrice
0   1          60       RL    ...           WD         Normal    208500
1   2          20       RL    ...           WD         Normal    181500
2   3          60       RL    ...           WD         Normal    223500
3   4          70       RL    ...           WD        Abnorml    140000
4   5          60       RL    ...           WD         Normal    250000

[5 rows x 81 columns]

In [6]:
home_data.describe()

Id   MSSubClass      ...             YrSold      SalePrice
count  1460.000000  1460.000000      ...        1460.000000    1460.000000
mean    730.500000    56.897260      ...        2007.815753  180921.195890
std     421.610009    42.300571      ...           1.328095   79442.502883
min       1.000000    20.000000      ...        2006.000000   34900.000000
25%     365.750000    20.000000      ...        2007.000000  129975.000000
50%     730.500000    50.000000      ...        2008.000000  163000.000000
75%    1095.250000    70.000000      ...        2009.000000  214000.000000
max    1460.000000   190.000000      ...        2010.000000  755000.000000

[8 rows x 38 columns]

In [7]:
test_data_columns = test_data.columns.tolist()
print(f"Columns in test data: {test_data_columns}")

Columns in test data: ['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating', 'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual', 'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond', 'PavedDrive', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch'

In [8]:
# create new features for trial 1: 
t1_features = []

# add old features
for f in features:
    if f in test_data_columns:
        t1_features.append(f)
    
# new features
for new_f in ['OverallQual', 'OverallCond', 'GrLivArea', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd']:
    if new_f not in t1_features and new_f in test_data_columns:
        t1_features.append(new)
        
print(f"Trial 1 features: {t1_features}")

NameError: name 'new' is not defined

In [9]:
# trial 1 data
t1_X = home_data[t1_features]
t1_X.head()

LotArea  YearBuilt      ...       BedroomAbvGr  TotRmsAbvGrd
0     8450       2003      ...                  3             8
1     9600       1976      ...                  3             6
2    11250       2001      ...                  3             6
3     9550       1915      ...                  3             7
4    14260       2000      ...                  4             9

[5 rows x 7 columns]

In [10]:
t1_X.describe()

LotArea    YearBuilt      ...       BedroomAbvGr  TotRmsAbvGrd
count    1460.000000  1460.000000      ...        1460.000000   1460.000000
mean    10516.828082  1971.267808      ...           2.866438      6.517808
std      9981.264932    30.202904      ...           0.815778      1.625393
min      1300.000000  1872.000000      ...           0.000000      2.000000
25%      7553.500000  1954.000000      ...           2.000000      5.000000
50%      9478.500000  1973.000000      ...           3.000000      6.000000
75%     11601.500000  2000.000000      ...           3.000000      7.000000
max    215245.000000  2010.000000      ...           8.000000     14.000000

[8 rows x 7 columns]

In [11]:
t1_X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 7 columns):
LotArea         1460 non-null int64
YearBuilt       1460 non-null int64
1stFlrSF        1460 non-null int64
2ndFlrSF        1460 non-null int64
FullBath        1460 non-null int64
BedroomAbvGr    1460 non-null int64
TotRmsAbvGrd    1460 non-null int64
dtypes: int64(7)
memory usage: 79.9 KB


In [12]:
# creat new model
t1_rf_model = RandomForestRegressor(random_state=1)

# fit new model
t1_rf_model.fit(t1_X, y)

/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
           oob_score=False, random_state=1, verbose=0, warm_start=False)

In [13]:
# trial 1 test data
t1_test_X = test_data[t1_features]
t1_test_X.describe()

LotArea    YearBuilt      ...       BedroomAbvGr  TotRmsAbvGrd
count   1459.000000  1459.000000      ...        1459.000000   1459.000000
mean    9819.161069  1971.357779      ...           2.854010      6.385195
std     4955.517327    30.390071      ...           0.829788      1.508895
min     1470.000000  1879.000000      ...           0.000000      3.000000
25%     7391.000000  1953.000000      ...           2.000000      5.000000
50%     9399.000000  1973.000000      ...           3.000000      6.000000
75%    11517.500000  2001.000000      ...           3.000000      7.000000
max    56600.000000  2010.000000      ...           6.000000     15.000000

[8 rows x 7 columns]

In [14]:
t1_test_X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1459 entries, 0 to 1458
Data columns (total 7 columns):
LotArea         1459 non-null int64
YearBuilt       1459 non-null int64
1stFlrSF        1459 non-null int64
2ndFlrSF        1459 non-null int64
FullBath        1459 non-null int64
BedroomAbvGr    1459 non-null int64
TotRmsAbvGrd    1459 non-null int64
dtypes: int64(7)
memory usage: 79.9 KB


In [15]:
# make predictions
t1_preds = t1_rf_model.predict(t1_test_X)

#output = pd.DataFrame({'Id': test_data.Id,
#                       'SalePrice': t1_preds})
#output.to_csv('submission2.csv', index=False)

# Improve Model: Trial 2
trial 2 uses trial 1 data and feature and tries to find best number of leaf nodes; for testing uses trial 1 test data

In [16]:
# creat a function to fit the model and calculate the MAE
def rf_calc_mea(max_leaf_nodes, train_X, val_X, train_y, val_y):
    # create random forest (rf) model
    rf_model = RandomForestRegressor(max_leaf_nodes=max_leaf_nodes, random_state=1)
    
    # fit the model
    rf_model.fit(train_X, train_y)
    
    # predict
    pred_y = rf_model.predict(val_X)
    
    # calculate MAE
    return mean_absolute_error(val_y, pred_y)
    
    

In [17]:
# list of possible number of leaf nodes
possible_leaf_nodes = range(100,401,10)

# split data
train_X, val_X, train_y, val_y = train_test_split(t1_X, y, random_state=1)

# list to save MAEs calculated
all_mae = []

# calculate MAEs
for leaf_nodes in possible_leaf_nodes:
    all_mae.append(rf_calc_mea(max_leaf_nodes=leaf_nodes, train_X=train_X, val_X=val_X, train_y=train_y, val_y=val_y))

/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The de

In [18]:
print(all_mae)
best_tree_index = all_mae.index(min(all_mae))
best_tree_size = possible_leaf_nodes[best_tree_index]
print(f"\nThe best tree size is {best_tree_size}.")

[22838.460825927963, 22857.26580070669, 22901.9662261218, 22822.962037040143, 22733.563312982733, 22780.827668509202, 22800.388298038783, 22885.999847504205, 22882.254009411343, 22826.545382815242, 22844.85236543475, 22732.146639592782, 22675.299406643942, 22676.153584465977, 22670.78175689906, 22646.233721306868, 22613.798382636713, 22637.144900814066, 22648.411385323885, 22580.13325161881, 22630.14273390463, 22601.845330450233, 22583.343534558415, 22585.76540410046, 22619.836743111373, 22621.138532614284, 22635.175696546885, 22645.831405897145, 22656.18204795752, 22645.580493406007, 22674.543902290123]

The best tree size is 290.


In [19]:
# fit t1_X assigning the max_leaf_nodes option
# creat new model
t2_rf_model = RandomForestRegressor(max_leaf_nodes=best_tree_size, random_state=1)

# fit new model
t2_rf_model.fit(t1_X, y)

/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=290,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
           oob_score=False, random_state=1, verbose=0, warm_start=False)

In [20]:
# make predictions
t2_preds = t2_rf_model.predict(t1_test_X)

#output = pd.DataFrame({'Id': test_data.Id,
#                       'SalePrice': t2_preds})
#output.to_csv('submission3.csv', index=False)

## Trial 2 (submission3) resulted in a worse MAE than Trial 1 (submission2)
### *Therefore, fitting the model without tuning resulted in better predictions*

# Improve Model: Trial 3

In [21]:
home_data.head()

Id  MSSubClass MSZoning    ...     SaleType  SaleCondition SalePrice
0   1          60       RL    ...           WD         Normal    208500
1   2          20       RL    ...           WD         Normal    181500
2   3          60       RL    ...           WD         Normal    223500
3   4          70       RL    ...           WD        Abnorml    140000
4   5          60       RL    ...           WD         Normal    250000

[5 rows x 81 columns]

In [22]:
# add 2 new columns to the original data
home_data['YearsOld'] = 2019 - home_data['YearBuilt']
home_data['YearsRemod'] = 2019 - home_data['YearRemodAdd']
home_data.head(10)

Id  MSSubClass MSZoning    ...      SalePrice  YearsOld YearsRemod
0   1          60       RL    ...         208500        16         16
1   2          20       RL    ...         181500        43         43
2   3          60       RL    ...         223500        18         17
3   4          70       RL    ...         140000       104         49
4   5          60       RL    ...         250000        19         19
5   6          50       RL    ...         143000        26         24
6   7          20       RL    ...         307000        15         14
7   8          60       RL    ...         200000        46         46
8   9          50       RM    ...         129900        88         69
9  10         190       RL    ...         118000        80         69

[10 rows x 83 columns]

In [23]:
# add the same columns that are added to home_data to the test_data
test_data['YearsOld'] = 2019 - test_data['YearBuilt']
test_data['YearsRemod'] = 2019 - test_data['YearRemodAdd']
test_data.head(10)

Id  MSSubClass MSZoning    ...      SaleCondition  YearsOld YearsRemod
0  1461          20       RH    ...             Normal        58         58
1  1462          20       RL    ...             Normal        61         61
2  1463          60       RL    ...             Normal        22         21
3  1464          60       RL    ...             Normal        21         21
4  1465         120       RL    ...             Normal        27         27
5  1466          60       RL    ...             Normal        26         25
6  1467          20       RL    ...             Normal        27         12
7  1468          60       RL    ...             Normal        21         21
8  1469          20       RL    ...             Normal        29         29
9  1470          20       RL    ...             Normal        49         49

[10 rows x 82 columns]

In [24]:
test_data_columns = test_data.columns.tolist()
print(test_data_columns)

['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating', 'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual', 'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond', 'PavedDrive', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'PoolQC'

In [25]:
# features to add
to_add_features = ['MSSubClass', 'LotFrontage', 'YearsOld', 'YearsRemod', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'TotalBsmtSF',
                  'LowQualFinSF', 'BsmtFullBath', 'BsmtHalfBath', 'HalfBath', 'Fireplaces', 'WoodDeckSF', 'OpenPorchSF',
                  'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'MiscVal', 'GarageCars', 'GarageArea']
t3_features = []

for f in to_add_features:
    if f not in t1_features and f in test_data_columns:
        t3_features.append(f)

# features to remove
remove_features = ['YearBuilt']
for r_f in remove_features:
    t3_features.pop(t3_features.index(r_f))

print(t3_features)

ValueError: 'YearBuilt' is not in list

In [26]:
t3_data = home_data[t3_features]
t3_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 22 columns):
MSSubClass       1460 non-null int64
LotFrontage      1201 non-null float64
YearsOld         1460 non-null int64
YearsRemod       1460 non-null int64
MasVnrArea       1452 non-null float64
BsmtFinSF1       1460 non-null int64
BsmtFinSF2       1460 non-null int64
TotalBsmtSF      1460 non-null int64
LowQualFinSF     1460 non-null int64
BsmtFullBath     1460 non-null int64
BsmtHalfBath     1460 non-null int64
HalfBath         1460 non-null int64
Fireplaces       1460 non-null int64
WoodDeckSF       1460 non-null int64
OpenPorchSF      1460 non-null int64
EnclosedPorch    1460 non-null int64
3SsnPorch        1460 non-null int64
ScreenPorch      1460 non-null int64
PoolArea         1460 non-null int64
MiscVal          1460 non-null int64
GarageCars       1460 non-null int64
GarageArea       1460 non-null int64
dtypes: float64(2), int64(20)
memory usage: 251.0 KB


### Notice that t3_data have missing values. Therefore, remove rows that has null values from the data.

In [27]:
# t3 colums are the features of t3 and the prediction variable 'SalePrice'
t3_colums = [f for f in t3_features]
t3_colums.append('SalePrice')
print(t3_colums)

['MSSubClass', 'LotFrontage', 'YearsOld', 'YearsRemod', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'TotalBsmtSF', 'LowQualFinSF', 'BsmtFullBath', 'BsmtHalfBath', 'HalfBath', 'Fireplaces', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'MiscVal', 'GarageCars', 'GarageArea', 'SalePrice']


In [28]:
t3_data = home_data[t3_colums].dropna(axis=0)
t3_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1195 entries, 0 to 1459
Data columns (total 23 columns):
MSSubClass       1195 non-null int64
LotFrontage      1195 non-null float64
YearsOld         1195 non-null int64
YearsRemod       1195 non-null int64
MasVnrArea       1195 non-null float64
BsmtFinSF1       1195 non-null int64
BsmtFinSF2       1195 non-null int64
TotalBsmtSF      1195 non-null int64
LowQualFinSF     1195 non-null int64
BsmtFullBath     1195 non-null int64
BsmtHalfBath     1195 non-null int64
HalfBath         1195 non-null int64
Fireplaces       1195 non-null int64
WoodDeckSF       1195 non-null int64
OpenPorchSF      1195 non-null int64
EnclosedPorch    1195 non-null int64
3SsnPorch        1195 non-null int64
ScreenPorch      1195 non-null int64
PoolArea         1195 non-null int64
MiscVal          1195 non-null int64
GarageCars       1195 non-null int64
GarageArea       1195 non-null int64
SalePrice        1195 non-null int64
dtypes: float64(2), int64(21)
memory u

In [29]:
# t3 training and prediction data
t3_X = t3_data[t3_features]
t3_y = t3_data['SalePrice']

In [30]:
t3_test_colums = [f for f in t3_data if f != 'SalePrice']
t3_test_colums.append('Id')

# remove nulls from test_data too
t3_test_data = test_data[t3_test_colums].dropna(axis=0)
t3_test_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1218 entries, 0 to 1458
Data columns (total 23 columns):
MSSubClass       1218 non-null int64
LotFrontage      1218 non-null float64
YearsOld         1218 non-null int64
YearsRemod       1218 non-null int64
MasVnrArea       1218 non-null float64
BsmtFinSF1       1218 non-null float64
BsmtFinSF2       1218 non-null float64
TotalBsmtSF      1218 non-null float64
LowQualFinSF     1218 non-null int64
BsmtFullBath     1218 non-null float64
BsmtHalfBath     1218 non-null float64
HalfBath         1218 non-null int64
Fireplaces       1218 non-null int64
WoodDeckSF       1218 non-null int64
OpenPorchSF      1218 non-null int64
EnclosedPorch    1218 non-null int64
3SsnPorch        1218 non-null int64
ScreenPorch      1218 non-null int64
PoolArea         1218 non-null int64
MiscVal          1218 non-null int64
GarageCars       1218 non-null float64
GarageArea       1218 non-null float64
Id               1218 non-null int64
dtypes: float64(9), int6

In [31]:
t3_test_X = t3_test_data[t3_features]

In [32]:
# creat t3 model
t3_rf_model = RandomForestRegressor(random_state=1)

# fit the model
t3_rf_model.fit(t3_X, t3_y)

/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
           oob_score=False, random_state=1, verbose=0, warm_start=False)

In [33]:
t3_preds = t3_rf_model.predict(t3_test_X)

output = pd.DataFrame({'Id': t3_test_data.Id,
                       'SalePrice': t3_preds})
output.to_csv('submission4.csv', index=False)

# *Failed* to submit Trial 3 (submission4) because the competition expects the data to have *1459 rows*!!!

# Test Your Work
After filling in the code above:
1. Click the **Commit and Run** button. 
2. After your code has finished running, click the small double brackets **<<** in the upper left of your screen.  This brings you into view mode of the same page. You will need to scroll down to get back to these instructions.
3. Go to the output tab at top of your screen. Select the button to submit your file to the competition.  
4. If you want to keep working to improve your model, select the edit button. Then you can change your model and repeat the process.

Congratulations, you've started competing in Machine Learning competitions.

# Continuing Your Progress
There are many ways to improve your model, and **experimenting is a great way to learn at this point.**

The best way to improve your model is to add features.  Look at the list of columns and think about what might affect home prices.  Some features will cause errors because of issues like missing values or non-numeric data types. 

Level 2 of this micro-course will teach you how to handle these types of features. You will also learn to use **xgboost**, a technique giving even better accuracy than Random Forest.


# Other Micro-Courses
The **[Pandas Micro-Course](https://kaggle.com/Learn/Pandas)** will give you the data manipulation skills to quickly go from conceptual idea to implementation in your data science projects. 

You are also ready for the **[Deep Learning](https://kaggle.com/Learn/Deep-Learning)** micro-course, where you will build models with better-than-human level performance at computer vision tasks.

---
**[Machine Learning Micro-Course Home Page](https://www.kaggle.com/learn/machine-learning)**

